### PR TITLE
ci: replace flake8 by ruff

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

it shall be a speed improvement
plus ruff seems to integrate much more fixes which we can enable later on:
```
#    "D",  # see: https://pypi.org/project/pydocstyle
#    "N",  # see: https://pypi.org/project/pep8-naming
#    "C4",  # see: https://pypi.org/project/flake8-comprehensions
#    "PT",  # see: https://pypi.org/project/flake8-pytest-style
#    "RET",  # see: https://pypi.org/project/flake8-return
#    "SIM",  # see: https://pypi.org/project/flake8-simplify
```
addresses https://github.com/Lightning-AI/lightning/pull/16430#issuecomment-1396503148

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @carmocca @akihironitta @borda

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -104,8 +104,12 @@ repos:
...
+    rev: 'v0.0.226'
     hooks:
-      - id: flake8
-        name: Check PEP8
+      - id: ruff
+        # Respect `exclude` and `extend-exclude` settings.
+        args:
+          - "--fix"
+          - "--respect-gitignore"
+          - "--force-exclude"
```
Comment (carmocca) [inline .pre-commit-config.yaml:115]: Why do we need this one?

```suggestion
```

File: pyproject.toml [hunk 2]
```diff
@@ -24,6 +43,57 @@ line-length = 120
...
+    ".git",
+    ".mypy_cache",
+    ".ruff_cache",
+    "__pypackages__",
+    "_build",
+    "build",
+    "dist",
+    "docs",
+    "_notebooks"
+]
```
Comment (carmocca) [inline pyproject.toml:77]: Can we use `respect-gitignore` instead of this?

File: pyproject.toml [hunk 3]
```diff
@@ -24,6 +43,57 @@ line-length = 120
...
+# TODO
+#    "D",  # see: https://pypi.org/project/pydocstyle
+#    "N",  # see: https://pypi.org/project/pep8-naming
+]
+#extend-select = [
+#    "C4",  # see: https://pypi.org/project/flake8-comprehensions
+#    "PT",  # see: https://pypi.org/project/flake8-pytest-style
+#    "RET",  # see: https://pypi.org/project/flake8-return
+#    "SIM",  # see: https://pypi.org/project/flake8-simplify
+#]
```
Comment (carmocca) [inline pyproject.toml:61]: How did you choose these? We didn't have them before

Comment (Borda) [reply pyproject.toml:61]: yes, we did not that is why they are commented but would experiment with then in the following work...

File: pyproject.toml (hunk 2 above)
Comment (Borda) [reply pyproject.toml:77]: not all are in git ignore such as docs, but we can prune it...

File: .pre-commit-config.yaml (hunk 1 above)
Comment (Borda) [reply .pre-commit-config.yaml:115]: it is mentioned in ruff docs/example to follow wxlude but it is now much more explained... we can test it without

File: pyproject.toml (hunk 3 above)
Comment (carmocca) [reply pyproject.toml:61]: Then I suggest you remove these from this PR and add them separately on those followups

**Review by justusschock — APPROVED** [review]

File: pyproject.toml (hunk 3 above)
Comment (Borda) [inline pyproject.toml:61]:
```suggestion
]
```

File: pyproject.toml [hunk 4]
```diff
@@ -24,6 +43,48 @@ line-length = 120
...
+    "dist",
+    "docs",
+    "_notebooks"
+]
+ignore-init-module-imports = true
+
+[tool.ruff.per-file-ignores]
+"setup.py" = ["D100", "SIM115"]
+"__about__.py" = ["D100"]
+"__init__.py" = ["D100"]
```
Comment (Borda) [inline pyproject.toml:74]:
```suggestion
```

File: pyproject.toml [hunk 5]
```diff
@@ -24,6 +43,48 @@ line-length = 120
...
+
+[tool.ruff.pydocstyle]
+# Use Google-style docstrings.
+convention = "google"
+
+#[tool.ruff.pycodestyle]
+#ignore-overlong-task-comments = true
+
```
Comment (Borda) [inline pyproject.toml:82]:
```suggestion
```

File: .pre-commit-config.yaml (hunk 1 above)
Comment (Borda) [reply .pre-commit-config.yaml:114]: tested and this needs to stay

**Review by carmocca — APPROVED** [review]

**Pull request merged**